dra-evolution/pkg/api/capacity_types.go

```diff
@@ -69,28 +74,94 @@ type ResourcePoolSpec struct {
 	// them) empty pool.
 }
 
+const ResourcePoolMaxSharedCapacity = 128
 const ResourcePoolMaxDevices = 128
 
+type DeviceShape struct {
+	// Attributes defines the attributes of this device shape.
+	// The name of each attribute must be unique.
+	//
+	// Must not have more than 32 entries.
+	//
+	// +listType=atomic
+	// +optional
+	Attributes []DeviceAttribute `json:"attributes,omitempty" protobuf:"bytes,3,opt,name=attributes"`
+
+	// Partitions defines the set of partitions into which this device shape
+	// may be allocated. If not populated, then the device shape is always
+	// consumed in its entirety.
+	//
+	// +listType=atomic
+	// +optional
+	Partitions []DevicePartition `json:"partitions,omitempty"`
+
+	// SharedCapacity defines the set of shared capacity consumable by
+	// partitions in this DeviceShape. Not meaninful for non-partitioned
+	// devices.
+	//
+	// Must not have more than 128 entries.
+	//
+	// +listType=atomic
+	// +optional
+	SharedCapacity []SharedCapacity `json:"sharedCapacity,omitempty"`
+}
+
 // Device represents one individual hardware instance that can be selected based
 // on its attributes.
+type DevicePartition struct {
+	// Name is unique identifier among all partitions for this device. The
+	// device name as recorded in the allocation will be the concatenation
+	// of the device name and the partition name with a '-' separator.
+	//
+	// NOTE: may need a better naming scheme
+	//
+	// It must be a DNS label.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// Attributes defines the attributes of this partition.
+	// The name of each attribute must be unique. The values
+	// in here are overlayed on top of the values in the device
+	// shape (overwriting them if the names are the same).
+	//
+	// NOTE: probably can get away with fewer
+	//
+	// Must not have more than 32 entries.
+	//
+	// +listType=atomic
+	// +optional
+	Attributes []DeviceAttribute `json:"attributes,omitempty" protobuf:"bytes,3,opt,name=attributes"`
+
+	// SharedCapacityConsumed defines the set of shared capacity consumed by
+	// this partition.
+	//
+	// Must not have more than 32 entries.
+	//
+	// +listType=atomic
+	// +optional
+	SharedCapacityConsumed []SharedCapacity `json:"sharedCapacityConsumed,omitempty"`
```

Comment:
Am I limited to the `SharedCapacity` from the `DeviceShape` where this partition is embedded? So, for example, if there were some "uncore" resources shared by all GPUs we couldn't consume that now?

Reply:
In the canonical example, the `DeviceShape` is a physical GPU, with a map of how it is partitonable. The `Devices` then just list the physical GPUs. I am not clear on what shared resources might live *outside* the GPU. Are you saying you need to capture shared resources that are not part of the GPU and somehow span GPUs?

Reply:
Yes, that is what I'm saying. I don't necessarily see it for GPUs, but I'm sure there are other devices that will want the ability to do this.

Reply:
Then we can add something later for that...